### PR TITLE
ECP keypair utility functions

### PR DESCRIPTION
**Pull request by gilles-peskine-arm**
Add utility functions to the ECP module to query and construct key pairs and points. This replaces the direct access to `mbedtls_ecp_keypair` and `mbedtls_ecp_point` fields that was possible in Mbed TLS 2.x. Fixes https://github.com/Mbed-TLS/mbedtls/issues/5441, fixes https://github.com/Mbed-TLS/mbedtls/issues/8367, fixes https://github.com/Mbed-TLS/mbedtls/issues/8652.

Update the sample programs. They now no longer do any direct access to private fields of `mbedtls_ecp_keypair` or `mbedtls_ecp_point`. Fixes https://github.com/Mbed-TLS/mbedtls/issues/5017. Note that any unrelated improvement of sample programs or their testing is out of scope here. I used the following to test them:
* run `programs/pkey/ecdsa`
* run `smoke.sh secp256r1 && smoke.sh x25519` using [smoke.sh.txt](https://github.com/Mbed-TLS/mbedtls/files/11840530/smoke.sh.txt)

Follow-up: use the new functions to simplify the [PSA transition guide](https://github.com/Mbed-TLS/mbedtls/pull/7766).

## PR checklist

Please tick as appropriate and edit the reasons (e.g.: "backport: not needed because this is a new feature")

- [x] **changelog** provided
- [x] **backport** no (new feature)
- [x] **tests** provided

### CODE REVIEW

File: tests/suites/test_suite_ecp.function
Comment (gilles-peskine-arm) [inline tests/suites/test_suite_ecp.function:1017]: If I understand https://jenkins-mbedtls.oss.arm.com/job/mbed-tls-pr-head/job/PR-7815-head/3/execution/node/7001/log/ correctly, outcome analysis complains that the `ecp_calc_public` tests are not executed when all EC algorithms use PSA. But I'd expect that, since the test function depends on `MBEDTLS_ECP_C`. Am I misunderstanding the test report? What did I do wrong?

Comment (mpg) [reply tests/suites/test_suite_ecp.function:1017]: The way parity testing is currently done, you need to add the description of the tests that are expected to be skipped to [https://github.com/Mbed-TLS/mbedtls/blob/development/tests/scripts/analyze_outcomes.py#L213](this list) in `analyze_outcomes.py`.

Comment (gilles-peskine-arm) [reply tests/suites/test_suite_ecp.function:1017]: Wow, that's pretty clumsy. I see there's a mechanism to ignore whole test suites, and individual test cases, but not test functions. Are you going to add one? Or should we split test suites into our-implementation-only stuff vs also-with-driver stuff?

**Comment by tom-cosgrove-arm** [comment]
This increases code size by 200 bytes when compiled for TF-M (their config, P-256 only, no `programs/` used) with ArmClang 6.19 with `-Oz`

| orig:5647d06be | pr:7815:1c8f9d61d |      Delta | Filename |
|---------------:|-----------:|-----------:|:---------|
|    6464, 0 |    6664, 0 |     +200+0 | `ecp.o` |

File: include/mbedtls/ecp.h
Comment (mpg) [reply include/mbedtls/ecp.h:786]: I was indeed going to argue it's frivolous. I won't fully object to adding an exporter based on this reason - after all, if the exporter is not used, it's going to be garbage-collected by the linker, but my personal preference would be changing the example programs.

I'm slightly inclined to remove "show-off" output from programs, but more importantly, if there's output that's not "show-off" (like, exporting a public key), we should use a standard format for it. This might uncover that we're missing  missing functions for actual use cases.

IMO the risk of adding a generic exporter that also works for "show-off" output is we'll be tempted to use it for cases where an actual application would have to export in a real usable format. (Like, you generated an `ecdsa_keypair` and want to export the public key - you can "show off" that using the coordinates exporter, but in the real world you'd want to export it in a real format that you can `pk_parse()` or `psa_import()` later, and I'm not sure we have a good story for that right now. Edit: well, not this particular one, but this PR is also adding a function for that. But you get the idea.)

File: ChangeLog.d/ecp-keypair-utilities.txt
Comment (mpg) [inline ChangeLog.d/ecp-keypair-utilities.txt:2]: Minor: I think it would help for the ChangeLog to name the functions that were added, in order to make it easier for users to discover them.

File: include/mbedtls/ecp.h
Comment (mpg) [inline include/mbedtls/ecp.h:1294]: Might want to mention `mbedtls_ecp_keypair_calc_public()` here.

File: tests/suites/test_suite_ecp.function
Comment (mpg) [reply tests/suites/test_suite_ecp.function:1017]: In the meantime we've added regex support to exclude a family of test cases with similar names. I don't think ignoring based on test function would be doable without some big changes, as the analysis is based on the outcomes files doesn't record the test function unless I'm mistaken.

If the list becomes too clumsy, splitting the test suite is indeed a good idea. In this case I don't think the split would be our-implementation-only vs also-with-drivers: it would rather be `ECP_LIGHT` vs "the rest of `ECP_C`".

**Review by mpg — CHANGES_REQUESTED** [review]
Only a design review for now. I'm mostly happy except for `export_coordinates()` where I disagree with the current API and am not convinced we really need it at all (but not opposed to adding it either).

File: include/mbedtls/ecp.h
Comment (gilles-peskine-arm) [reply include/mbedtls/ecp.h:786]: The ECP interface allows users to build their own cryptographic primitives on top of the core arithmetic. I would expect ECDSA, ECDH and ECJPAKE to only use public ECP functions, since users should be able to write similar things on their own. If not with `mbedtls_ecp_point_export_coordinates`, how could those modules use the result of an operation that calculates a point?

Comment (mpg) [reply include/mbedtls/ecp.h:786]: Indeed if we consider it a goal for ECP to support people building their own crypto on top, then we need some form of `export_coordinates`.

In my mind that was never really a goal: ECP is for us to build the protocols we implement, and `ecp.h` is only public because that's how we did things back then. But since we never said otherwise (and in particular did not make `ecp.h` internal in 3.0), it's of course perfectly reasonable for people to consider it public, and then the API should be usable not just by us.

File: ChangeLog.d/ecp-keypair-utilities.txt
Comment (daverodgman) [inline ChangeLog.d/ecp-keypair-utilities.txt:3]: Also fixes #8367 and #5017

**Comment by gilles-peskine-arm** [comment]
I've force-pushed an update that's rebased on top of development and has new commits to address the comments so far, apart from those related to `mbedtls_ecp_point_export_coordinates`. I'll make another update for that once the CI passes, either by amending the documentation or the function or by removing the function (possibly by rewriting the history to never add the function).

The previous version is in https://github.com/gilles-peskine-arm/mbedtls/tree/ecp-export-partial-2.

File: include/mbedtls/ecp.h
Comment (gilles-peskine-arm) [reply include/mbedtls/ecp.h]: Having thought it over,

> I'm really not happy about some public's function documentation saying "in the internal representation used by Mbed TLS"

is a decisive argument. Neither am I. Code that really wants this (to build efficient arithmetic for a custom ECC-based algorithm?) can access private fields. Code that just wants to obtain a concrete representation of the point can use `mbedtls_ecp_point_write_binary`. So I've removed `mbedtls_ecp_point_export_coordinates`.

**Comment by gilles-peskine-arm** [comment]
The history is now a bit messy due to a back-and-forth in the design. If this is too annoying to review, I could simplify it to never add `mbedtls_ecp_point_export_coordinates`.

**Comment by gilles-peskine-arm** [comment]
I've pushed a new history that never creates `mbedtls_ecp_point_export_coordinates`. The actual development history, with that function added and removed and with some commits to fix minor issues based on CI or review feedback, but rebased midway on top of a recent `development` to avoid merge conflicts, is in https://github.com/gilles-peskine-arm/mbedtls/tree/ecp-export-partial-4.

File: tests/suites/test_suite_ecp.data
Comment (mpg) [inline tests/suites/test_suite_ecp.data:632]: Out of curiosity: why the 3 cases of public after private with different groups? I feel I'd only have included the first one, and you probably have a reason for including more.

File: tests/suites/test_suite_ecp.function
Comment (mpg) [inline tests/suites/test_suite_ecp.function:1126]: Hmm, it seems to me this entire test file is only compiled when `MBEDTLS_ECP_LIGHT` is enabled, and then that means so is `BIGNUM_C`. Am I missing something? What did you have in mind?

(If you're thinking about driver-only builds, those that have `BIGNUM_C` disabled simply don't compile the functions in this file.)

(More occurrences below, I won't point each of them.)

File: tests/suites/test_suite_ecp.function
Comment (mpg) [inline tests/suites/test_suite_ecp.function:1404]: Minor: I'd find it more readable with three paragraphs like:
```c
        mbedtls_ecp_group_free(&export_grp);
        mbedtls_ecp_group_init(&export_grp);
        TEST_EQUAL(mbedtls_ecp_export(&key, &export_grp, NULL, NULL), 0);
        TEST_EQUAL(mbedtls_ecp_group_cmp(&key.grp, &export_grp), 0);
```
With the current ordering, it was not immediately clear to me what was happening here.

**Review by mpg — CHANGES_REQUESTED** [review]
Looking pretty good, just a few minor things and a couple of questions.

File: include/mbedtls/ecp.h
Comment (gilles-peskine-arm) [reply include/mbedtls/ecp.h:1437]: I don't think we should. It's reasonable that if the public/private part is missing from the key then `*Q`/`*d` is set to an unspecified value.

File: programs/pkey/gen_key.c
Comment (gilles-peskine-arm) [reply programs/pkey/gen_key.c:199]: I agree, we're inconsistent, but that's preexisting and out of scope.

File: tests/suites/test_suite_ecp.data
Comment (gilles-peskine-arm) [reply tests/suites/test_suite_ecp.data:632]: Knowing how our implementation works, it makes sense to test with at least one NIST curve, at least one Koblitz curve and at least one Montgomery curve. Maybe it doesn't matter for this specific function, but that feels like a good default rule (short of insisting on testing every function with every curve).

File: tests/suites/test_suite_ecp.function
Comment (gilles-peskine-arm) [reply tests/suites/test_suite_ecp.function:1126]: I don't remember why I put guards in this function. To be robust with respect to what ECP light might have become, maybe? (I wrote this code back in June 2023.) Or to work with `MBEDTLS_ECP_ALT`? Anyway, the guards aren't bad, but they are noise unless we generalize the test suite, which, six months later, isn't going to happen. Should I remove those guards?

Comment (mpg) [reply tests/suites/test_suite_ecp.function:1126]: `ECP_ALT` doesn't change anything here: the only parts of `ECP_C` it affects are outside of `ECP_LIGHT`, which does require `BIGNUM_C` and will continue doing so for the foreseeable future.

So IMO these guards are noise that should be removed, to avoid giving readers a false impression.

**Review by mpg — COMMENTED** [review]
Looking good, apart from the superfluous `BIGNUM_C` guards that I'd like removed.

**Comment by mpg** [comment]
I'm removing "needs-design-approval" as I think we know have an agreement on the design.

**Review by mpg — APPROVED** [review]
LGTM

**Review by minosgalanakis — COMMENTED** [review]
Looks good overall. 

Two notes:

- I assume that the setter for public key `mbedtls_ecp_read_public_key`  is intentionally missing? 

  If a user needs to importa  public key they should import the private one using    `mbedtls_ecp_read_key ()` and then calculate the public using `mbedtls_ecp_keypair_calc_public()` ?

- The `show_ecp_key(const mbedtls_ecp_keypair *ecp, int has_private)` method could be either a helper function for programs or even an API call? Given the complexity to produce a readable printout, we should at least point the users to how to do it. 

The latter is not a strong ask, and should not block this PR.

**Comment by gilles-peskine-arm** [comment]
> I assume that the setter for public key mbedtls_ecp_read_public_key is intentionally missing?

I'm not sure what that would be. How would it differ from `mbedtls_ecp_set_public_key`?

> The show_ecp_key(const mbedtls_ecp_keypair *ecp, int has_private) method could be either a helper function for programs

It could be, but those programs are currently self-contained. They're sample programs, so there's a strong argument that people should be able to just take the one `.c` file and modify it. That's a high barrier for introducing a new dependency. Conversely, this is code that we rarely modify, so the cost of duplication is low. So this is a rare case where I think code duplication is ok, even if we didn't have time constraints.

> or even an API call? Given the complexity to produce a readable printout, we should at least point the users to how to do it.

Printing out keys in a readable format is rarely done apart from demos. So it doesn't belong in the API, and I don't think we need to call it out in documentation.

**Review by minosgalanakis — APPROVED** [review]
LGTM

**Pull request merged**
